### PR TITLE
feat(username): System property for enforcing username

### PR DESCRIPTION
**Pull request by aeng**
https://zanata.atlassian.net/browse/ZNTA-855

---
This change is [Reviewable](https://reviewable.io/reviews/zanata/zanata-server/1173)

### CODE REVIEW

**Comment by seanf** [comment]
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[docs/release-notes.md, line 5 \[r1\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI64W9dBgVTFZZeDmk3:-KI64W9dBgVTFZZeDmk4:-1359203913) ([raw file](https://github.com/zanata/zanata-server/blob/b1a8c94b2cfce40ce4050012317f99d9e6ec1b65/docs/release-notes.md#L5)):*
> ```Markdown
> <h5>New feature</h5>
> * [ZNTA-746](https://zanata.atlassian.net/browse/ZNTA-746) - Add shorcut key for approve and reject translation
> * [ZNTA-855](https://zanata.atlassian.net/browse/ZNTA-855) - System property to enforce username for registration
> ```

That description (and the commit) should really start with "add system property..."

---

*[zanata-war/src/main/java/org/zanata/ApplicationConfiguration.java, line 133 \[r1\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI64oEB0rvZB-Offbt4:-KI64oEB0rvZB-Offbt5:-1673287759) ([raw file](https://github.com/zanata/zanata-server/blob/b1a8c94b2cfce40ce4050012317f99d9e6ec1b65/zanata-war/src/main/java/org/zanata/ApplicationConfiguration.java#L133)):*
> ```Java
>      * User are only allow to input username themselves when:
>      * - If no username is available from response
>      * - If username is conflicting with existing username
> ```

I think it would be better to  throw an exception in those two cases, because they both imply a serious configuration problem in the system. I think it is better to require a sysadmin to fix the login system, or to SQL UPDATE the previous incorrect username, than to allow mismatched usernames when the sysadmin has told Zanata to enforce matching usernames.

---

*[zanata-war/src/main/java/org/zanata/action/NewProfileAction.java, line 99 \[r1\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI66HxuBV-X_CE5TnWU:-KI66HxuBV-X_CE5TnWV:1346338494) ([raw file](https://github.com/zanata/zanata-server/blob/b1a8c94b2cfce40ce4050012317f99d9e6ec1b65/zanata-war/src/main/java/org/zanata/action/NewProfileAction.java#L99)):*
> ```Java
>         return applicationConfiguration.isEnforceUsername() &&
>             StringUtils.isNotBlank(username) && isUsernameValid(username) &&
>             !isUsernameTaken(username);
> ```

As above. "Enforce" should mean "enforce".

---

*[zanata-war/src/main/java/org/zanata/action/NewProfileAction.java, line 109 \[r1\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI65sFbyJIq8G37cObI:-KI65sFcTfORDxvth0bC:-286916786) ([raw file](https://github.com/zanata/zanata-server/blob/b1a8c94b2cfce40ce4050012317f99d9e6ec1b65/zanata-war/src/main/java/org/zanata/action/NewProfileAction.java#L109)):*
> ```Java
>      */
>     private boolean isUsernameValid(String username) {
>         Pattern p = Pattern.compile("^[a-z\\d_]{3,20}$");
> ```

Where did this magic regex appear from? Can't we share a constant or something?

---

*[zanata-war/src/main/webapp/profile/create_user.xhtml, line 60 \[r1\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI665sTeVKzcGQljSSg:-KI665sTeVKzcGQljSSh:570545836) ([raw file](https://github.com/zanata/zanata-server/blob/b1a8c94b2cfce40ce4050012317f99d9e6ec1b65/zanata-war/src/main/webapp/profile/create_user.xhtml#L60)):*
> ```HTML
>             <zanata:decorate id="username" componentId="usernameField">
>               <ui:fragment rendered="#{not newProfileAction.readOnlyUsername}">
>                 <h:inputText id="username" required="true" maxlength="20"
> ```

Instead of hard-coding the length 20, we should share a constant with the other places this length is used.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1173)*

**Comment by seanf** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[zanata-war/src/main/java/org/zanata/ApplicationConfiguration.java, line 143 \[r1\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI681IoYGIfN6HXt9Go:-KI681IpfBnXlb2-Nxmw:1696234041) ([raw file](https://github.com/zanata/zanata-server/blob/b1a8c94b2cfce40ce4050012317f99d9e6ec1b65/zanata-war/src/main/java/org/zanata/ApplicationConfiguration.java#L143)):*
> ```Java
>      */
>     @Getter
>     private boolean enforceUsername;
> ```

I think a more meaningful name would be better, perhaps zanata.enforceMatchingUsernames or zanata.usernamesMustMatchExternal.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1173)*

**Comment by aeng** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[zanata-war/src/main/java/org/zanata/action/NewProfileAction.java, line 109 \[r1\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI65sFbyJIq8G37cObI:-KI68V0tItT1pHHfqyJ1:-2011402620) ([raw file](https://github.com/zanata/zanata-server/blob/b1a8c94b2cfce40ce4050012317f99d9e6ec1b65/zanata-war/src/main/java/org/zanata/action/NewProfileAction.java#L109)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> Where did this magic regex appear from? Can't we share a constant or something?
</details>
See {@link #getUsername()}

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1173)*

**Comment by seanf** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[zanata-war/src/main/java/org/zanata/action/NewProfileAction.java, line 109 \[r1\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI65sFbyJIq8G37cObI:-KI695qA0vAkPLYyIxjG:1871115391) ([raw file](https://github.com/zanata/zanata-server/blob/b1a8c94b2cfce40ce4050012317f99d9e6ec1b65/zanata-war/src/main/java/org/zanata/action/NewProfileAction.java#L109)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary>
> See {@link #getUsername()}
</details>
We should share a constant with org.zanata.action.AbstractProfileAction#getUsername() then.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1173)*

**Comment by aeng** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[zanata-war/src/main/java/org/zanata/ApplicationConfiguration.java, line 133 \[r1\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI64oEB0rvZB-Offbt4:-KI6M_qmwz-s7yVabwFb:-499146009) ([raw file](https://github.com/zanata/zanata-server/blob/b1a8c94b2cfce40ce4050012317f99d9e6ec1b65/zanata-war/src/main/java/org/zanata/ApplicationConfiguration.java#L133)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> I think it would be better to  throw an exception in those two cases, because they both imply a serious configuration problem in the system. I think it is better to require a sysadmin to fix the login system, or to SQL UPDATE the previous incorrect username, than to allow mismatched usernames when the sysadmin has told Zanata to enforce matching usernames.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1173)*

**Comment by seanf** [comment]
Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[zanata-war/src/main/java/org/zanata/action/AbstractProfileAction.java, line 23 \[r2\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI6VoBu0TI5zxCoEF7d:-KI6VoBu0TI5zxCoEF7e:-60837418) ([raw file](https://github.com/zanata/zanata-server/blob/5fcd544340232a5af746980367287269b035b429/zanata-war/src/main/java/org/zanata/action/AbstractProfileAction.java#L23)):*
> ```Java
> public abstract class AbstractProfileAction {
>     protected final static String USERNAME_REGEX = "^[a-z\\d_]{3,20}$";
>     protected final static int USERNAME_MAX_LENGTH = 20;
> ```

Public would be good for these constants.

---

*[zanata-war/src/main/java/org/zanata/action/NewProfileAction.java, line 98 \[r2\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI6WHBVWTfTdtrT3l0D:-KI6WHBWuOxiJCTCkpP0:-1385980717) ([raw file](https://github.com/zanata/zanata-server/blob/5fcd544340232a5af746980367287269b035b429/zanata-war/src/main/java/org/zanata/action/NewProfileAction.java#L98)):*
> ```Java
>                 } else if (!isUsernameValid(username)) {
>                     throw new ZanataServiceException(
>                         "System is set to use username matched with openId response but username is not in valid format." +
> ```

This new feature isn't restricted to OpenID is it? What about Kerberos setups? These error messages should probably talk about "external authentication" instead of "OpenID".

---

*[zanata-war/src/main/webapp/profile/create_user.xhtml, line 60 \[r2\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI6VRxouFWsS-LUVokz:-KI6VRxouFWsS-LUVol-:-2001872940) ([raw file](https://github.com/zanata/zanata-server/blob/5fcd544340232a5af746980367287269b035b429/zanata-war/src/main/webapp/profile/create_user.xhtml#L60)):*
> ```HTML
>             <zanata:decorate id="username" componentId="usernameField">
>               <ui:fragment rendered="#{not newProfileAction.readOnlyUsername}">
>                 <h:inputText id="username" required="true" maxlength="#{newProfileAction.usernameMaxLength}"
> ```

Apparently it will be possible to access the constant AbstractProfileAction.USERNAME_MAX_LENGTH: directly once we move to Java EE 7 http://stackoverflow.com/a/3735006/14379

We can't use that yet, but in the meantime, we might as well prepare by making that constant public.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1173)*

**Comment by seanf** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[zanata-war/src/main/java/org/zanata/action/NewProfileAction.java, line 98 \[r3\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI6ceGfFIzDFu9KGI5f:-KI6ceGfFIzDFu9KGI5g:155802229) ([raw file](https://github.com/zanata/zanata-server/blob/d29145d433b8c9485a0f49c6b9d1f89c84a4d8ee/zanata-war/src/main/java/org/zanata/action/NewProfileAction.java#L98)):*
> ```Java
>                 } else if (!isUsernameValid(username)) {
>                     throw new ZanataServiceException(
>                         "System is set to use username matched with external authentication response but username is not in valid format." +
> ```

To make these situations a little easier to diagnose, I would suggest "Server option enforceMatchingUsernames is set, but username from external authentication is (missing|already in use|not valid for Zanata)"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1173)*

**Comment by aeng** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[zanata-war/src/main/java/org/zanata/action/NewProfileAction.java, line 98 \[r2\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI6WHBVWTfTdtrT3l0D:-KI6fCvLeNL0mNcEOHDQ:-632720176) ([raw file](https://github.com/zanata/zanata-server/blob/5fcd544340232a5af746980367287269b035b429/zanata-war/src/main/java/org/zanata/action/NewProfileAction.java#L98)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> This new feature isn't restricted to OpenID is it? What about Kerberos setups? These error messages should probably talk about "external authentication" instead of "OpenID".
</details>
I don't get it, isn't a specific issue is easier for diagnose the issue with the username rather than 1 combined message?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1173)*

**Comment by aeng** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[zanata-war/src/main/java/org/zanata/action/NewProfileAction.java, line 98 \[r3\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI6ceGfFIzDFu9KGI5f:-KI6fRxUBkorTmHC4hEb:-632720176) ([raw file](https://github.com/zanata/zanata-server/blob/d29145d433b8c9485a0f49c6b9d1f89c84a4d8ee/zanata-war/src/main/java/org/zanata/action/NewProfileAction.java#L98)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> To make these situations a little easier to diagnose, I would suggest "Server option enforceMatchingUsernames is set, but username from external authentication is (missing|already in use|not valid for Zanata)"
</details>
I don't get it, isn't a specific issue is easier for diagnose the issue with the username rather than 1 combined message?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1173)*

**Comment by aeng** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[docs/release-notes.md, line 5 \[r1\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI64W9dBgVTFZZeDmk3:-KI6fZECywLau7rog4TA:-499146009) ([raw file](https://github.com/zanata/zanata-server/blob/b1a8c94b2cfce40ce4050012317f99d9e6ec1b65/docs/release-notes.md#L5)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> That description (and the commit) should really start with "add system property..."
</details>
Done.

---

*[zanata-war/src/main/java/org/zanata/ApplicationConfiguration.java, line 143 \[r1\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI681IoYGIfN6HXt9Go:-KI6f_PCJXRCj6WkCvMF:-499146009) ([raw file](https://github.com/zanata/zanata-server/blob/b1a8c94b2cfce40ce4050012317f99d9e6ec1b65/zanata-war/src/main/java/org/zanata/ApplicationConfiguration.java#L143)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> I think a more meaningful name would be better, perhaps zanata.enforceMatchingUsernames or zanata.usernamesMustMatchExternal.
</details>
Done.

---

*[zanata-war/src/main/java/org/zanata/action/AbstractProfileAction.java, line 23 \[r2\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI6VoBu0TI5zxCoEF7d:-KI6famkakuwCtGLTcPH:-499146009) ([raw file](https://github.com/zanata/zanata-server/blob/5fcd544340232a5af746980367287269b035b429/zanata-war/src/main/java/org/zanata/action/AbstractProfileAction.java#L23)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> Public would be good for these constants.
</details>
Done.

---

*[zanata-war/src/main/java/org/zanata/action/NewProfileAction.java, line 99 \[r1\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI66HxuBV-X_CE5TnWU:-KI6fW47hK3g7gM_hZ_e:-499146009) ([raw file](https://github.com/zanata/zanata-server/blob/b1a8c94b2cfce40ce4050012317f99d9e6ec1b65/zanata-war/src/main/java/org/zanata/action/NewProfileAction.java#L99)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> As above. "Enforce" should mean "enforce".
</details>
Done.

---

*[zanata-war/src/main/java/org/zanata/action/NewProfileAction.java, line 109 \[r1\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI65sFbyJIq8G37cObI:-KI6fVPjiNHJGbAT0Il3:-499146009) ([raw file](https://github.com/zanata/zanata-server/blob/b1a8c94b2cfce40ce4050012317f99d9e6ec1b65/zanata-war/src/main/java/org/zanata/action/NewProfileAction.java#L109)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> We should share a constant with org.zanata.action.AbstractProfileAction#getUsername() then.
</details>
Done.

---

*[zanata-war/src/main/webapp/profile/create_user.xhtml, line 60 \[r1\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI665sTeVKzcGQljSSg:-KI6fYPqlECt54MqS8xW:-499146009) ([raw file](https://github.com/zanata/zanata-server/blob/b1a8c94b2cfce40ce4050012317f99d9e6ec1b65/zanata-war/src/main/webapp/profile/create_user.xhtml#L60)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> Instead of hard-coding the length 20, we should share a constant with the other places this length is used.
</details>
Done.

---

*[zanata-war/src/main/webapp/profile/create_user.xhtml, line 60 \[r2\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI6VRxouFWsS-LUVokz:-KI6fX5ygHChisHh1DMg:-499146009) ([raw file](https://github.com/zanata/zanata-server/blob/5fcd544340232a5af746980367287269b035b429/zanata-war/src/main/webapp/profile/create_user.xhtml#L60)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> Apparently it will be possible to access the constant AbstractProfileAction.USERNAME_MAX_LENGTH: directly once we move to Java EE 7 http://stackoverflow.com/a/3735006/14379
> 
> We can't use that yet, but in the meantime, we might as well prepare by making that constant public.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1173)*

**Comment by seanf** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[zanata-war/src/main/java/org/zanata/action/NewProfileAction.java, line 98 \[r3\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI6ceGfFIzDFu9KGI5f:-KI6iXtpz9j5GHRj3oL_:-1986627587) ([raw file](https://github.com/zanata/zanata-server/blob/d29145d433b8c9485a0f49c6b9d1f89c84a4d8ee/zanata-war/src/main/java/org/zanata/action/NewProfileAction.java#L98)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary>
> I don't get it, isn't a specific issue is easier for diagnose the issue with the username rather than 1 combined message?
</details>
I'm not suggesting a single message with the literal string "(missing|already in use|not valid for Zanata)", I'm just suggesting different wording for all three messages, wording which includes the name of the system option in question.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1173)*

**Comment by aeng** [comment]
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[zanata-war/src/main/java/org/zanata/action/NewProfileAction.java, line 98 \[r3\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KI6ceGfFIzDFu9KGI5f:-KI6kbDIGrCJFP_FPrXq:-499146009) ([raw file](https://github.com/zanata/zanata-server/blob/d29145d433b8c9485a0f49c6b9d1f89c84a4d8ee/zanata-war/src/main/java/org/zanata/action/NewProfileAction.java#L98)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary>
> I'm not suggesting a single message with the literal string "(missing|already in use|not valid for Zanata)", I'm just suggesting different wording for all three messages, wording which includes the name of the system option in question.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1173)*

**Comment by seanf** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1173)*

**Comment by seanf** [comment]
@aeng this has some merge conflicts...

<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary>
> #### feat\(username\): System property for enforcing username
> 
> https://zanata.atlassian.net/browse/ZNTA-855
> 
> 
> 
</details>


---

Reviewed 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[docs/release-notes.md, line 24 \[r5\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KIkthTgTtrSwyu5U6tJ:-KIkthTgTtrSwyu5U6tK:-65392012) ([raw file](https://github.com/zanata/zanata-server/blob/b6b81591ac5312e459ebcc0fb137630f1c250687/docs/release-notes.md#L24)):*
> ```Markdown
> 
> <h5>Infrastructure Changes</h5>
> * System admin can set system property `zanata.enforce.matchingusernames` to enforce matching username to be used for new user registration.
> ```

I suppose this should be `zanata.enforce.matching.usernames` now, if we're not going to use camelCase. What do you think @carlosmunoz?

---

*[zanata-war/src/main/java/org/zanata/action/NewProfileAction.java, line 95 \[r5\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KIl5oSZDWlvMZf13Swt:-KIl5oSZDWlvMZf13Swu:1088486837) ([raw file](https://github.com/zanata/zanata-server/blob/b6b81591ac5312e459ebcc0fb137630f1c250687/zanata-war/src/main/java/org/zanata/action/NewProfileAction.java#L95)):*
> ```Java
>             //validate username if enforcing matching username is enabled
>             if (applicationConfiguration.isEnforceMatchingUsernames()) {
>                 if(StringUtils.isBlank(username) || isUsernameTaken(username) || !isUsernameValid(username)) {
> ```

I'm not keen on the double-checking of these conditions: too hard to maintain. I liked the old if-else chain better, even if you have to duplicate the line facesMessage.addGlobal line.

---

*[zanata-war/src/main/java/org/zanata/action/NewProfileAction.java, line 117 \[r5\]](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-KIl69K9aIQD49N7ngTr:-KIl69K9aIQD49N7ngTs:890939947) ([raw file](https://github.com/zanata/zanata-server/blob/b6b81591ac5312e459ebcc0fb137630f1c250687/zanata-war/src/main/java/org/zanata/action/NewProfileAction.java#L117)):*
> ```Java
>                     }
>                     facesMessages.addGlobal(SEVERITY_ERROR, message);
>                     throw new ZanataServiceException(exceptionMessage);
> ```

Do we really need to add a global faces message AND throw an exception? Won't the exception be caught and shown to the user anyway?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-server/1173#-:-KIl8VAu3JRSU9JKQEt_:507333473)*

**Comment by djansen-redhat** [comment]
@aeng Is this missing a documentation snippet, for the new property?

**Comment by aeng** [comment]
@djansen-redhat I believe it's in release notes. Any other place you think I should include the information?

**Comment by djansen-redhat** [comment]
@aeng maybe in docs/user-guide/system-admin/configuration/authentication.md?

**Comment by djansen-redhat** [comment]
- [ ] Username exists: red error screen with "Login Failed. But you can help us fix it!"
```
12:06:31,453 ERROR [org.zanata.exception.handler.AbstractExceptionHandler] (default task-6) exception happened in view: /profile/create_user.xhtml: org.zanata.exception.ZanataServiceException: Server option zanata.enforce.matchingusernames is set, but username from external authentication is already in use. djansentest
	at org.zanata.action.NewProfileAction.onCreate(NewProfileAction.java:117)
```

**Comment by aeng** [comment]
@djansen-redhat Does the page have some other message? It's the correct landing page, but should have other messages.

File: zanata-war/src/main/java/org/zanata/action/NewProfileAction.java [hunk 1]
```diff
@@ -92,29 +92,18 @@ public void onCreate() {
...
-                                        + USERNAME_REGEX;
-                    }
-                    facesMessages.addGlobal(SEVERITY_ERROR, message);
-                    throw new ZanataServiceException(exceptionMessage);
+                if (StringUtils.isBlank(username)) {
+                    throw new ZanataServiceException(
+                            "Server option zanata.enforce.matchingusernames is set, but username from external authentication is missing.");
+                } else if (isUsernameTaken(username)) {
+                    throw new ZanataServiceException(
+                            "Server option zanata.enforce.matchingusernames is set, but username from external authentication is already in use."
```
Comment (djansen-redhat) [inline zanata-war/src/main/java/org/zanata/action/NewProfileAction.java]: Missing a space after "use.". Also, should this be a colon, i.e. "... already in use: "?

**Comment by djansen-redhat** [comment]
:white_check_mark: tested

**Pull request merged**